# Hotel Booking Demand Dataset - Exploratory Data Analysis

## Day 1: Understanding and Profiling

# Project Objective

The objective of this project is to analyze hotel booking data to understand customer booking behavior, hotel demand patterns, pricing trends, stay preferences, and booking cancellations.

The analysis aims to transform raw booking data into meaningful insights that can support better decision-making for hotel management and help customers understand booking trends and demand patterns.

# Business Understanding

The hotel industry faces challenges such as fluctuating demand, booking cancellations, seasonal variations, and changing customer preferences.

This project focuses on analyzing hotel booking data from Resort Hotels and City Hotels to identify customer behavior patterns, booking trends, pricing behavior, and cancellation factors.

The insights generated from this analysis can help hotels improve planning, pricing strategies, resource allocation, and customer experience while also helping travelers make informed booking decisions.

# Dataset Description

Dataset Name: Hotel Booking Demand Dataset

Dataset Summary:

- Total Rows: 119,390
- Total Columns: 32
- Each row represents a hotel booking.
- The dataset contains booking information from Resort Hotels and City Hotels.
- The dataset includes numerical, categorical, and date-related features.

Key information available includes hotel type, booking details, customer information, stay duration, room details, pricing information, and reservation status.

In [18]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

### Use Case

Import libraries required for data loading, analysis, and visualization.

### Findings, Interpretation & Business Insight

- Pandas will be used for data manipulation and analysis.
- NumPy will support numerical operations.
- Matplotlib and Seaborn will be used for data visualization.
- These libraries provide the foundation required for performing Exploratory Data Analysis.

### Overall Insight

The analytical environment has been prepared successfully for data exploration and visualization.

In [19]:
df = pd.read_csv("../data/raw/hotel_bookings.csv")

In [20]:
df.head()

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,304.0,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,NaN,0,Transient,98.0,0,1,Check-Out,2015-07-03


### Purpose

The `head()` function is used to display the first few records of the dataset and verify that the data has been loaded correctly. It provides an initial understanding of the available features and the structure of the dataset.

### Key Observations

- Each row represents a hotel booking record.
- The dataset contains information related to hotel type, booking details, customer characteristics, stay duration, pricing, and reservation status.
- The first few records belong to the **Resort Hotel** category.
- Features such as `lead_time`, `arrival_date`, `adults`, `adr`, and `reservation_status` are available for analysis.
- Missing values can already be observed in columns such as `agent` and `company`.
- The dataset contains a mix of numerical and categorical variables.

### Key Insight

The initial preview indicates that the dataset is comprehensive and contains sufficient information to analyze hotel demand, customer behavior, booking patterns, pricing trends, and cancellation outcomes. The presence of some missing values suggests that data preprocessing will be required in later stages.

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

### Purpose

The `info()` function is used to understand the dataset structure, data types, and identify missing values.

### Key Observations

- The dataset contains 119,390 records and 32 columns.
- It includes both numerical and categorical features.
- Missing values are present in `country`, `agent`, `company`, and `children`.
- Most columns contain complete information.

### Insight

The dataset is well-structured and suitable for exploratory analysis. Missing values identified in a few columns will be addressed during the data preprocessing stage.

In [23]:
df.describe()

,is_canceled,lead_time,arrival_date_year,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,children,babies,is_repeated_guest,previous_cancellations,previous_bookings_not_canceled,booking_changes,agent,company,days_in_waiting_list,adr,required_car_parking_spaces,total_of_special_requests
count,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,119386.000000,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,103050.000000,6797.000000,119390.000000,119390.000000,119390.000000,119390.000000
mean,0.370416,104.011416,2016.156554,27.165173,15.798241,0.927599,2.500302,1.856403,0.103890,0.007949,0.031912,0.087118,0.137097,0.221124,86.693382,189.266735,2.321149,101.831122,0.062518,0.571363
std,0.482918,106.863097,0.707476,13.605138,8.780829,0.998613,1.908286,0.579261,0.398561,0.097436,0.175767,0.844336,1.497437,0.652306,110.774548,131.655015,17.594721,50.535790,0.245291,0.792798
min,0.000000,0.000000,2015.000000,1.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,6.000000,0.000000,-6.380000,0.000000,0.000000
25%,0.000000,18.000000,2016.000000,16.000000,8.000000,0.000000,1.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,9.000000,62.000000,0.000000,69.290000,0.000000,0.000000
50%,0.000000,69.000000,2016.000000,28.000000,16.000000,1.000000,2.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,14.000000,179.000000,0.000000,94.575000,0.000000,0.000000
75%,1.000000,160.000000,2017.000000,38.000000,23.000000,2.000000,3.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,229.000000,270.000000,0.000000,126.000000,0.000000,1.000000
max,1.000000,737.000000,2017.000000,53.000000,31.000000,19.000000,50.000000,55.000000,10.000000,10.000000,1.000000,26.000000,72.000000,21.000000,535.000000,543.000000,391.000000,5400.000000,8.000000,5.000000


### Purpose

The `describe()` function provides a statistical summary of numerical features. It helps understand the distribution, central tendency, spread, and variability of the data.

### Key Observations

- The average booking lead time is approximately **104 days**, indicating that customers generally book hotels well in advance.
- The average **ADR (Average Daily Rate)** is around **102**, representing the typical room price paid per booking.
- The average number of **adults per booking is close to 2**, suggesting that couples form a significant portion of hotel guests.
- Most bookings contain **no children or babies**, as indicated by the median values of zero.
- The majority of guests stay around **2–3 week nights** and **1 weekend night** on average.
- The cancellation rate is approximately **37%**, indicating that cancellations represent a significant business challenge.
- Several features show a large difference between minimum and maximum values, suggesting substantial variation in customer behavior and booking patterns.
- Some variables such as ADR, lead time, waiting days, and previous bookings contain extreme values that may require outlier analysis.

### Key Insight

The statistical summary reveals diverse booking behaviors, stay patterns, and pricing characteristics. Customers generally book several months in advance, travel mostly as couples, and exhibit varying booking habits. The presence of extreme values and high variability highlights the importance of further analysis and preprocessing before drawing final conclusions.

In [24]:
df.nunique()

hotel                                2
is_canceled                          2
lead_time                          479
arrival_date_year                    3
arrival_date_month                  12
arrival_date_week_number            53
arrival_date_day_of_month           31
stays_in_weekend_nights             17
stays_in_week_nights                35
adults                              14
children                             5
babies                               5
meal                                 5
country                            177
market_segment                       8
distribution_channel                 5
is_repeated_guest                    2
previous_cancellations              15
previous_bookings_not_canceled      73
reserved_room_type                  10
assigned_room_type                  12
booking_changes                     21
deposit_type                         3
agent                              333
company                            352
days_in_waiting_list     

### Purpose

The `nunique()` function is used to determine the number of distinct values present in each feature. This helps understand feature diversity, identify categorical variables, and assess the uniqueness of information available in the dataset.

### Key Observations

- The dataset contains bookings from **2 hotel types**.
- The target variable **is_canceled** has **2 unique values**, representing canceled and non-canceled bookings.
- The **country** feature contains **177 unique values**, indicating a highly diverse international customer base.
- Features such as **market_segment**, **distribution_channel**, and **customer_type** contain a limited number of categories, making them suitable for customer segmentation analysis.
- **ADR** has **8,879 unique values**, showing substantial variation in room pricing across bookings.
- **Agent** and **company** contain a large number of unique identifiers, suggesting multiple booking sources and business partners.
- **reservation_status_date** contains **926 unique dates**, enabling time-based and trend analysis.

### Key Insight

The dataset contains a combination of low-cardinality and high-cardinality features. Low-cardinality variables are useful for grouping and comparison, while high-cardinality variables provide detailed information about customer behavior, booking sources, pricing patterns, and reservation activity.

In [30]:
numerical_cols = df.select_dtypes(include=['int64', 'float64']).columns.tolist()

print(numerical_cols)

['is_canceled', 'lead_time', 'arrival_date_year', 'arrival_date_week_number', 'arrival_date_day_of_month', 'stays_in_weekend_nights', 'stays_in_week_nights', 'adults', 'children', 'babies', 'is_repeated_guest', 'previous_cancellations', 'previous_bookings_not_canceled', 'booking_changes', 'agent', 'company', 'days_in_waiting_list', 'adr', 'required_car_parking_spaces', 'total_of_special_requests']


### Observation

The dataset contains 20 numerical features related to booking behavior, guest information, pricing, reservation history, and hotel operations.

Key numerical variables include lead time, ADR, number of guests, stay duration, previous cancellations, and special requests.

### Insight

These features will be useful for statistical analysis, distribution analysis, correlation studies, outlier detection, and understanding customer booking behavior.

In [31]:
categorical_cols = df.select_dtypes(include=['object']).columns.tolist()

print(categorical_cols)

['hotel', 'arrival_date_month', 'meal', 'country', 'market_segment', 'distribution_channel', 'reserved_room_type', 'assigned_room_type', 'deposit_type', 'customer_type', 'reservation_status', 'reservation_status_date']


### Observation

The dataset contains 12 categorical features, including hotel type, meal plan, country, market segment, distribution channel, customer type, and reservation status.

These variables describe customer characteristics, booking sources, and reservation outcomes.

### Insight

Categorical features will help analyze customer segments, booking channels, hotel preferences, and cancellation behavior across different groups.

In [32]:
date_columns = [
    'arrival_date_year',
    'arrival_date_month',
    'arrival_date_week_number',
    'arrival_date_day_of_month',
    'reservation_status_date'
]

print(date_columns)

['arrival_date_year', 'arrival_date_month', 'arrival_date_week_number', 'arrival_date_day_of_month', 'reservation_status_date']


### Observation

The dataset contains date-related features such as arrival year, month, week number, day of month, and reservation status date.

These variables capture the timing of bookings and reservation activities.

### Insight

Date features can be used to identify seasonal trends, peak booking periods, monthly demand patterns, and changes in customer behavior over time.

In [33]:
df.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

In [34]:
missing_values = df.isnull().sum()

missing_values[missing_values > 0]

children         4
country        488
agent        16340
company     112593
dtype: int64

### Purpose

The `isnull().sum()` function is used to identify missing values in each column and assess data completeness before preprocessing.

### Key Observations

- Most columns contain complete information with no missing values.
- Missing values are present in:
  - `children` → 4 missing values
  - `country` → 488 missing values
  - `agent` → 16,340 missing values
  - `company` → 112,593 missing values
- The `company` column contains the highest number of missing values, while `children` has only a few missing records.

### Key Insight

The dataset is generally clean, with missing values concentrated in a small number of columns. These missing values will need to be handled during preprocessing to improve data quality and ensure reliable analysis results.

# Data Quality Issue Log

| Issue | Observation |
|---------|------------|
| Missing Values | Present in children, country, agent, and company |
| Numerical Features | Successfully identified |
| Categorical Features | Successfully identified |
| Date Features | Available for time-based analysis |
| Dataset Structure | Suitable for EDA |

# Initial Insights

1. The dataset contains detailed information about hotel bookings, customer demographics, pricing, and reservation outcomes.

2. Customers generally book hotels well in advance, indicating the importance of demand forecasting and revenue planning.

3. The dataset includes customers from a large number of countries, suggesting a diverse international customer base.

4. Missing values are concentrated in a few columns and will need to be addressed during preprocessing.

5. The dataset contains both numerical and categorical variables, making it suitable for statistical analysis, segmentation, and trend analysis.

6. Date-related features can be used to study seasonality, booking trends, and hotel demand patterns.

# Day 1 Summary

### Key Findings

- The dataset contains 119,390 hotel booking records and 32 features.
- Numerical, categorical, and date-related features were successfully identified.
- Missing values were found in a small number of columns.
- The dataset contains sufficient information to analyze customer behavior, booking patterns, hotel demand, pricing, and cancellations.
- Initial profiling confirms that the dataset is suitable for exploratory data analysis.

### Conclusion

Day 1 focused on understanding the dataset structure, feature characteristics, and data quality. Initial observations indicate that the dataset is well-structured and contains valuable information for hotel booking analysis. The next step is to clean and preprocess the data before performing deeper exploratory analysis.